mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/hook/ClientEventHookManagerImpl.java

```diff
@@ -0,0 +1,170 @@
+ */
+
+package org.apache.rocketmq.mqtt.cs.hook;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.handler.codec.mqtt.MqttFixedHeader;
+import io.netty.handler.codec.mqtt.MqttMessageType;
+import io.netty.handler.codec.mqtt.MqttPublishMessage;
+import io.netty.handler.codec.mqtt.MqttPublishVariableHeader;
+import io.netty.handler.codec.mqtt.MqttQoS;
+import org.apache.rocketmq.common.ThreadFactoryImpl;
+import org.apache.rocketmq.mqtt.common.hook.ClientEventHook;
+import org.apache.rocketmq.mqtt.common.hook.ClientEventHookManager;
+import org.apache.rocketmq.mqtt.common.hook.HookResult;
+import org.apache.rocketmq.mqtt.common.model.ClientEventMessage;
+import org.apache.rocketmq.mqtt.common.model.ClientEventType;
+import org.apache.rocketmq.mqtt.cs.channel.ChannelInfo;
+import org.apache.rocketmq.mqtt.cs.session.infly.MqttMsgId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.Resource;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.rocketmq.mqtt.common.model.Constants.CLIENT_EVENT_BATCH_SIZE;
+import static org.apache.rocketmq.mqtt.common.model.Constants.CLIENT_EVENT_ORIGIN_TOPIC;
+
+@Component
+public class ClientEventHookManagerImpl implements ClientEventHookManager {
+    @Resource
+    private MqttMsgId mqttMsgId;
+    private static Logger logger = LoggerFactory.getLogger(ClientEventHookManagerImpl.class);
+    private ClientEventHook clientEventHook;
+    private AtomicBoolean isAssembled = new AtomicBoolean(false);
+    private LinkedBlockingQueue<ClientEventMessage> eventQueue = new LinkedBlockingQueue<>(10000);
+    private ExecutorService executorService;
+
+    @PostConstruct
+    public void init() {
+        executorService = new ThreadPoolExecutor(
+                1,
+                1,
+                1,
+                TimeUnit.MINUTES,
+                new LinkedBlockingQueue<>(5000),
+                new ThreadFactoryImpl("ClientEventHookManager_"));
+        executorService.submit(new ClientEventHookExecution());
+    }
+
+    @Override
+    public void addHook(ClientEventHook clientEventHook) {
+        if (isAssembled.get()) {
+            throw new IllegalArgumentException("ClientEventHook Was Assembled");
+        }
+        this.clientEventHook = clientEventHook;
+        isAssembled.set(true);
+    }
+
+    private class ClientEventHookExecution implements Runnable {
```

Comment:
Why not take the initiative directly, but instead save batches? Online and offline events can be sent immediately, just like ordinary messages, and there seems to be no performance bottleneck.